src/pykeen/nn/emb.py

```diff
@@ -580,3 +581,246 @@ def forward(
+
+        # initialize
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Reset the model's parameters."""
+        for w in (
+            self.w_loop,
+            self.w_fwd,
+            self.w_bwd,
+            self.w_rel,
+            self.self_loop,
+        ):
+            nn.init.xavier_uniform_(w)
+        self.bias.reset_parameters()
+
+    def message(
+        self,
+        x_e: torch.FloatTensor,
+        x_r: torch.FloatTensor,
+        triples: Optional[Tuple[torch.LongTensor, torch.LongTensor, torch.LongTensor]],
+    ) -> torch.FloatTensor:
+        if triples is None:
+            # compose
+            m = self.composition(x_e, self.self_loop)
+            # transform (no dropout)
+            return m @ self.self_loop
+
+        # compose
+        h, r, t = triples
+        m = self.composition(x_e[h], x_r[r])
+        # transform
+        m = m @ self.w
+        # TODO: normalization
+        # self.in_norm = self.compute_norm(self.in_index, num_ent)
+        # aggregate by sum
+        x_e = torch.zeros_like(x_e).index_add(dim=0, index=t, source=m)
+        return self.drop(x_e)
+
+    def forward(
+        self,
+        x_e: torch.FloatTensor,
+        x_r: torch.FloatTensor,
+        triples: torch.LongTensor,
+    ) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+        # split
+        h, r, t = triples.t()
+
+        # entity transformation
+        x_e = sum((
+            self.message(
+                x_e=x_e,
+                x_r=x_r,
+                triples=triples_,
+            ),
+            for triples_ in (
+            (h, 2 * r, t),  # forward; todo make sure this is the correct interpretation of inverse relations
+            (t, 2 * r + 1, h),  # backward
+            None,  # self-loop
+        )
+        )) / 3
+        if self.bias:
+            x_e = self.bias(x_e)
+        x_e = self.bn(x_e)
+        x_e = self.activation(x_e)
+
+        # Relation transformation
+        x_r = self.w_rel(x_r)
+        return x_e, x_r
+
+
+class CompGCNRepresentation(nn.Module):
+    """A sequence of CompGCN layers."""
+
+    def __init__(
+        self,
+        *,
+        triples_factory: TriplesFactory,
+        embedding_specification: EmbeddingSpecification,
+        num_layers: int = 1,
+        output_dim: int = 128,
+        layer_kwargs: Optional[Mapping[str, Any]] = None,
+    ):
+        super().__init__()
+        # TODO: Check
+        assert not triples_factory.create_inverse_triples
+        self.entity_representations = embedding_specification.make(
+            num_embeddings=triples_factory.num_entities,
+        )
+        self.relation_representations = embedding_specification.make(
+            num_embeddings=2 * triples_factory.num_relations,
+        )
+        input_dim = self.entity_representations.embedding_dim
+        layers = []
+        for i in range(num_layers):
```

Comment:
i is unused. would there be any reason for a layer to know what position it's in?

Reply:
No, the layers process the input sequentially.